proof/refine/ARM/Tcb_R.thy

```diff
@@ -794,49 +796,541 @@ lemma threadSet_valid_objs_tcbPriority_update:
+  unfolding getEpQueue_def by wpsimp auto
+
+lemma updateEpQueue_triv: "ep \<noteq> IdleEP \<Longrightarrow> updateEpQueue ep (epQueue ep) = ep"
+  by (cases ep; clarsimp simp: updateEpQueue_def)
+
+lemma updateEPQueue_IdleEP[simp]: "(updateEpQueue ep qs = IdleEP) = (ep = IdleEP)"
+  by (cases ep; simp add: updateEpQueue_def)
 
-lemma sp_corres: "corres dc (einvs and tcb_at t) (invs' and tcb_at' t and valid_objs' and (\<lambda>_. x \<le> maxPriority))
-                     (set_priority t x) (setPriority t x)"
+lemma reorderEp_invs':
+  "\<lbrace>invs' and st_tcb_at' (\<lambda>st. epBlocked st = Some ntfnPtr) tptr\<rbrace>
+   reorderEp ntfnPtr tptr
+   \<lbrace>\<lambda>rv. invs'\<rbrace>"
+  apply (simp only: reorderEp_def)
+  apply (subst bind_assoc[symmetric, where m="tcbEPDequeue tptr _"])
+  apply (wp set_ep_minor_invs')
+       apply (simp add: pred_conj_def live_ntfn'_def)
+       apply (wpsimp wp: getEndpoint_wp tcbEPAppend_valid_ep' tcbEPAppend_rv_wf' tcbEPAppend_rv_wf''
+                         tcbEPDequeue_valid_ep' tcbEPDequeue_rv_wf' tcbEPDequeue_rv_wf'')+
+  apply (frule ep_ko_at_valid_objs_valid_ep', fastforce)
+  apply (clarsimp simp: updateEpQueue_triv sym_refs_asrt_def valid_ep'_def pred_tcb_at'_def
+                        obj_at'_def projectKO_eq projectKO_tcb projectKO_ep)
+  apply (frule_tac ko=obj and p=tptr in sym_refs_ko_atD'[rotated])
+   apply (clarsimp simp: obj_at'_def projectKO_eq projectKO_tcb)
+  apply (case_tac "tcbState obj"; clarsimp simp: epBlocked_def split: ntfn.splits if_splits)
+    apply (auto simp: invs'_def valid_state'_def valid_idle'_def if_live_then_nonz_cap'_def
+                      refs_of_rev' get_refs_def ko_wp_at'_def obj_at'_def projectKO_eq projectKO_tcb
+               split: option.splits)
+  done
+
+lemma threadSetPriority_invs':
+  "\<lbrace>invs' and obj_at' (Not \<circ> tcbQueued) t and K (p \<le> maxPriority)\<rbrace>
+   threadSetPriority t p
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  apply (rule hoare_gen_asm)
+  apply (wpsimp wp: threadSet_invs_trivial simp: threadSetPriority_def)
+  by (fastforce simp: tcb_cte_cases_def invs'_def valid_state'_def inQ_def obj_at'_def
+                      valid_queues_def valid_queues_no_bitmap_def valid_release_queue'_def)+
+
+lemma tcbSchedEnqueue_all_invs_but_ct_not_inQ:
+  "\<lbrace>all_invs_but_ct_not_inQ' and st_tcb_at' runnable' t\<rbrace>
+   tcbSchedEnqueue t
+   \<lbrace>\<lambda>_. all_invs_but_ct_not_inQ'\<rbrace>"
+  apply (simp add: invs'_def valid_state'_def)
+  apply (rule hoare_pre)
+   apply (wp tcbSchedEnqueue_ct_not_inQ valid_irq_node_lift irqs_masked_lift hoare_vcg_disj_lift
+             valid_irq_handlers_lift' cur_tcb_lift ct_idle_or_in_cur_domain'_lift2
+             untyped_ranges_zero_lift
+        | simp add: cteCaps_of_def o_def
+        | auto elim!: st_tcb_ex_cap'' valid_objs'_maxDomain valid_objs'_maxPriority
```

Comment:
Is the `auto` necessary here? I'm not sure whether this counts as using `auto` before the final command of a proof

Reply:
I think we might technically allow it if it's completely unavoidable, but it's definitely not the preferred style.
Out of curiosity, what goes wrong with the standard approach here?
```
apply (wpsimp ...)
apply (auto ...)

Reply:
This is actually the same proof as `tcbSchedEnqueue_invs'` in `Schedule_R`, I just copied it.